src/ZenohUTransport.cpp

```diff
@@ -10,3 +10,356 @@
+	return zenoh_key.str();
+}
+
+std::vector<std::pair<std::string, std::string>>
+ZenohUTransport::uattributesToAttachment(const UAttributes& attributes) {
+	std::vector<std::pair<std::string, std::string>> res;
+
+	std::string version(&UATTRIBUTE_VERSION, 1);
+
+	std::string data;
+	attributes.SerializeToString(&data);
+
+	res.push_back(std::make_pair("", version));
+	res.push_back(std::make_pair("", data));
+	return res;
+}
+
+UAttributes ZenohUTransport::attachmentToUAttributes(
+    const AttachmentView& attachment) {
+	std::vector<BytesView> attachment_vec;
+	attachment.iterate(
+	    [&](const BytesView& key, const BytesView& value) -> bool {
+		    attachment_vec.push_back(value);
+		    return true;
+	    });
+
+	if (attachment_vec.size() != 2) {
+		// TODO: error report, exception?
+	}
+
+	if (attachment_vec[0].get_len() == 1) {
+		if (attachment_vec[0].as_string_view()[0] != UATTRIBUTE_VERSION) {
+			// TODO: error report, exception?
+		}
+	};
+	UAttributes res;
+	// TODO: more efficient way?
+	res.ParseFromString(std::string(attachment_vec[1].as_string_view()));
+	return res;
+}
+
+Priority ZenohUTransport::mapZenohPriority(UPriority upriority) {
+	switch (upriority) {
+		case UPriority::UPRIORITY_CS0:
+			return Z_PRIORITY_BACKGROUND;
+		case UPriority::UPRIORITY_CS1:
+			return Z_PRIORITY_DATA_LOW;
+		case UPriority::UPRIORITY_CS2:
+			return Z_PRIORITY_DATA;
+		case UPriority::UPRIORITY_CS3:
+			return Z_PRIORITY_DATA_HIGH;
+		case UPriority::UPRIORITY_CS4:
+			return Z_PRIORITY_INTERACTIVE_LOW;
+		case UPriority::UPRIORITY_CS5:
+			return Z_PRIORITY_INTERACTIVE_HIGH;
+		case UPriority::UPRIORITY_CS6:
+			return Z_PRIORITY_REAL_TIME;
+		case UPriority::UPRIORITY_UNSPECIFIED:
+		default:
+			return Z_PRIORITY_DATA_LOW;
+	}
+}
+
+UMessage ZenohUTransport::sampleToUMessage(const Sample& sample) {
+	UMessage message;
+	if (sample.get_attachment().check()) {
+		*message.mutable_attributes() =
+		    attachmentToUAttributes(sample.get_attachment());
+	}
+	std::string payload(sample.get_payload().as_string_view());
+	message.set_payload(payload);
+
+	return message;
+}
+
+ZenohUTransport::ZenohUTransport(const UUri& defaultUri,
+                                 const std::filesystem::path& configFile)
+    : UTransport(defaultUri),
+      session_(expect<Session>(open(
+          std::move(expect(config_from_file(configFile.string().c_str())))))) {}
+
+UStatus ZenohUTransport::registerRequestListener_(const std::string& zenoh_key,
+                                                  CallableConn listener) {
+	// NOTE: listener is captured by copy here so that it does not go out
+	// of scope when this function returns.
+	auto on_query = [this, listener](const Query& query) {
+		UAttributes attributes;
+		if (query.get_attachment().check()) {
+			attributes = attachmentToUAttributes(query.get_attachment());
+		}
+		auto id_str = serializer::uuid::AsString().serialize(attributes.id());
+		std::unique_lock<std::mutex> lock(query_map_mutex_);
```

Comment:
Wouldn't it be better to take the data structs that need thread safety, and turn them into classes holding the map or whatever and the mutex, and having accesor methods to do the locking? This style would make the main code much more readable.

Reply:
Make sense, but before we should solve the question about cleanupListener signature, because if it will be as is (but not as rust one) we should rework these internals types.

Reply:
@sashacmc - I'll summarize what we discussed on `cleanupListener` over slack here so its available for everyone.

`cleanupListener()` is not a direct equivalent to `unregisterListener()` in other language libraries. Since we are using [connection objects](https://github.com/eclipse-uprotocol/up-cpp/blob/main/include/up-cpp/utils/CallbackConnection.h) to represent listener callbacks, by the time `cleanupListener()` is called the connection is already broken. Calling a disconnected handle from the transport side will do nothing.

The purpose of `cleanupListener()` is purely informational - it exists to inform the transport implementation that a connection to a listener / callback function has been broken and that, depending on the details of the transport's implementation, it _may_ need to schedule a cleanup operation. Since the connection to the callback function is already broken, this cleanup does not need to occur immediately. It can be deferred if that makes the cleanup easier.

Additionally, while the listener handle parameter passed to `cleanupListener()` _can_ be compared against other listener handles to find matching handles, that is not strictly necessary. The `CallerHandle` class used there has both an [`isConnected()` method](https://github.com/eclipse-uprotocol/up-cpp/blob/main/include/up-cpp/utils/CallbackConnection.h#L371) and a [boolean conversion operator](https://github.com/eclipse-uprotocol/up-cpp/blob/main/include/up-cpp/utils/CallbackConnection.h#L361) that will return false if the handle is no longer connected. A transport implementation _could_ just remove all unconnected handles after `cleanupListener()` is called.